specs/AppLifecycle/Restart/restartApi.markdown

```diff
@@ -0,0 +1,259 @@
+
+2.  **Restart-me-after-termination**. The Win32 API
+    [RegisterApplicationRestart](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-registerapplicationrestart)
+    enables an app to register itself to be restarted after termination,
+    and to provide an arbitrary command-line string for the restarted
+    instance. The reasons for termination include app crash or hang, app
+    update, or system update. There's also a matching
+    [UnregisterApplicationRestart](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-unregisterapplicationrestart)
+    API.
+
+There are 2 related Win32 APIs that focus on recovery prior to restart:
+
+1.  **Let-me-do-something-before-termination**. An app can call
+    [RegisterApplicationRecoveryCallback](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-registerapplicationrecoverycallback)
+    to register a callback for the system to call before terminating the
+    app. If an application encounters an unhandled exception or becomes
+    unresponsive, Windows Error Reporting (WER) calls the specified
+    recovery callback, where the app can save state information. The
+    system pings the app every n seconds to make sure that it hasn't
+    hung in its callback. The app can specify the ping interval when it
+    registers the callback. In its recovery callback, the app can call
+    RegisterApplicationRestart a second time, to update the
+    command-line.
+
+2.  **Recovery-in-progress**. While the app is doing work in its
+    recovery callback, the app must periodically call
+    [ApplicationRecoveryInProgress](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-applicationrecoveryinprogress)
+    \-- if it doesn't call within the registered ping interval, WER will
+    terminate the process.
+
+| App Type                                                            | Applicable Restart Mechanism                                           | Gap                                                                                                                  |
+| ------------------------------------------------------------------- | ---------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------- |
+| Win32 (unpackaged), Win32/Centennial (packaged), WinUI (Centennial) | RegisterApplicationRestart for OS + app update/restart/crash scenarios | Apps can register with the OS to restart in specific app/OS states, but cannot restart on-command in a healthy state |
+
+## Existing API Details
+
+<table>
+<colgroup>
+<col style="width: 41%" />
+<col style="width: 58%" />
+</colgroup>
+<thead>
+<tr class="header">
+<th>Existing API</th>
+<th>Details</th>
+</tr>
+</thead>
+<tbody>
+<tr class="odd">
+<td><p>HRESULT <strong>RegisterApplicationRestart</strong>(PCWSTR pwzCommandline, DWORD dwFlags);</p></td>
```

Comment:
I'm confused, this looks like a Win32 API, but the sample makes it appear to be a WinRT API?

Reply:
This is a Win32 API (this section outlines the existing APIs): https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-registerapplicationrestart

Reply:
DISCUSSION: It's OK, showing the MSDN content. Just name the API and link to MSDN with it.

Reply:
Added links